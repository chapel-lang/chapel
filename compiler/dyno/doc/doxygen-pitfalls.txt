COMMON PITFALLS WITH DOXYGEN (THAT BREAK THE DOCS BUILD)

Whenever you run into a Doxygen problem that breaks the docs build,
make sure to explain what you did to fix it here.

---
ERROR: "WARNING: cpp: identifier reference target not found
---

Doxygen will fail to resolve:

  - C++ standard library types (that are not builtin, e.g. `int64_t`)
  - Any `compiler/next` types that are generated by macros (for now)
  - C++11 "in-class member initializers" for enum fields

Just add these types to `$CHPL_HOME/doc/util/nitpick_ignore`.

An "in-class member initializer" for a field looks like:

  `WhereClauseResult whereClauseResult = WHERE_TBD;`

Doxygen can't seem to resolve these when the RHS is an enum.

`base.rst` needs to include types that might be declared in the chpl
namespace elsewhere. Examples of this are chp::owned, chpl::Iterable,
and chpl::UniqueString.

---
"WARNING: Invalid C++ declaration: Expected end of definition."
---

Often followed by code like:

  `template<> MostSpecificCandidates >`
  `template<typename A, typename B> pair< A, B > >`

These are usually the definitions of `update()` functions that are
used by the query system. They are heavily templated. Just put the
function inside a DO_NOT_DOCUMENT block like so:

  /// \cond DO_NOT_DOCUMENT
  template<> struct update<resolution::ResolvedExpression> {
    bool operator()(resolution::ResolvedExpression& keep,
                    resolution::ResolvedExpression& addin) const {
      return defaultUpdate(keep, addin);
    }
  };
  /// \endcond

These probably aren't worth figuring out how to fix, and they don't
really need documentation anyway.

