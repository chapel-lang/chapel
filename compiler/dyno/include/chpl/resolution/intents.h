/*
 * Copyright 2021-2022 Hewlett Packard Enterprise Development LP
 * Other additional copyright holders may be indicated within.
 *
 * The entirety of this work is licensed under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 *
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#ifndef CHPL_RESOLUTION_INTENTS_H
#define CHPL_RESOLUTION_INTENTS_H

#include "chpl/types/QualifiedType.h"

namespace chpl {
namespace resolution {


/** Resolve the intent of a formal argument based on its type
    and (possibly generic) intent. For example 'const' is a generic
    intent and it has different behavior depending on the type.

    The argument 'isThis' indicates if the formal argument is
    the 'this' method receiver argument.

    The generic and resolved intents are all represented as QualifiedType::Kind.

    If the type is not known or only partially known, this function
    can return a generic intent.
 */
types::QualifiedType::Kind resolveIntent(const types::QualifiedType& t,
                                         bool isThis);


} // end namespace resolution
} // end namespace chpl

#endif
