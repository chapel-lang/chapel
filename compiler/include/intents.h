/*
 * Copyright 2020-2023 Hewlett Packard Enterprise Development LP
 * Copyright 2004-2019 Cray Inc.
 * Other additional copyright holders may be indicated within.
 *
 * The entirety of this work is licensed under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 *
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#ifndef INTENTS_H
#define INTENTS_H

// TODO: Consider moving other things (like printers) to this header
// when/if we want to clean things up further.

const int INTENT_FLAG_IN          = 0x01;
const int INTENT_FLAG_OUT         = 0x02;
const int INTENT_FLAG_CONST       = 0x04;
const int INTENT_FLAG_REF         = 0x08;
const int INTENT_FLAG_PARAM       = 0x10;
const int INTENT_FLAG_TYPE        = 0x20;
const int INTENT_FLAG_BLANK       = 0x40;
const int INTENT_FLAG_MAYBE_CONST = 0x80;

// If this enum is modified, ArgSymbol::intentDescrString()
// and intentDescrString(IntentTag) should also be updated to match
enum IntentTag {
  INTENT_IN              = INTENT_FLAG_IN,
  INTENT_OUT             = INTENT_FLAG_OUT,
  INTENT_INOUT           = INTENT_FLAG_IN          | INTENT_FLAG_OUT,
  INTENT_CONST           = INTENT_FLAG_CONST,
  INTENT_CONST_IN        = INTENT_FLAG_CONST       | INTENT_FLAG_IN,
  INTENT_REF             = INTENT_FLAG_REF,
  INTENT_CONST_REF       = INTENT_FLAG_CONST       | INTENT_FLAG_REF,
  INTENT_REF_MAYBE_CONST = INTENT_FLAG_MAYBE_CONST | INTENT_FLAG_REF,
  INTENT_PARAM           = INTENT_FLAG_PARAM,
  INTENT_TYPE            = INTENT_FLAG_TYPE,
  INTENT_BLANK           = INTENT_FLAG_BLANK
};

/*
enum ForallIntentTag : task- or forall-intent tags

TFI_IN_PARENT
  The compiler adds this shadow var during resolution for each TFI_IN
  and TFI_CONST_IN. A TFI_IN_PARENT represents the task function's formal
  that the corresponding TFI_IN or TFI_CONST_IN is to be initialized from.

TFI_REDUCE
  This shadow var replaces the uses of the outer variable in the loop body
  in case of a 'reduce' intent. This is done in parsing and scopeResolve.
  It is analogous to the TFI_IN shadow var for an 'in' intent.
  A TFI_REDUCE var represents the current task's accumulation state.

TFI_REDUCE_*
  The compiler adds one each of these shadow vars during resolution
  for each TFI_REDUCE. They represent:

  TFI_REDUCE_OP        - the current task's reduction OP
  TFI_REDUCE_PARENT_AS - the parent task's Accumulation State
  TFI_REDUCE_PARENT_OP - the parent task's reduction OP

  The *PARENT* vars, like TFI_IN_PARENT, are the task function's formals.

The remaining tags should be self-explanatory.
*/
enum ForallIntentTag {
  // user-specified intents
  TFI_DEFAULT,  // aka TFI_BLANK
  TFI_CONST,                       // ShadowVarSymbol nicknames:
  TFI_IN,                          //   SI
  TFI_CONST_IN,                    //   "
  TFI_REF,                         //   SR
  TFI_CONST_REF,                   //   "
  TFI_REDUCE,                      //   AS    (for Accumulation State)
  TFI_TASK_PRIVATE,                //   TPV
  // compiler-added helpers; note isCompilerAdded()
  TFI_IN_PARENT,                   //   INP
  TFI_REDUCE_OP,                   //   RP    (for Reduce oP)
  TFI_REDUCE_PARENT_AS,            //   PAS
  TFI_REDUCE_PARENT_OP,            //   PRP
};

const char* forallIntentTagDescription(ForallIntentTag tfiTag);

enum RetTag {
  RET_VALUE,
  RET_REF,
  RET_CONST_REF,
  RET_PARAM,
  RET_TYPE
};

#endif
