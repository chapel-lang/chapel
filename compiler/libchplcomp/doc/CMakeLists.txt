# Support for documentation of AST header
find_package(Doxygen)

if(DOXYGEN_FOUND)
  set(DOXYGEN_GENERATE_XML YES)
  set(DOXYGEN_OUTPUT_DIRECTORY "${CMAKE_CURRENT_BINARY_DIR}/doxygen")
  #file(MAKE_DIRECTORY ${DOXYGEN_OUTPUT_DIRECTORY})
  doxygen_add_docs(api-docs ${CHPL_MAIN_INCLUDE_DIR})
  #set(SPHINX_SOURCE ${CMAKE_CURRENT_SOURCE_DIR})
  #set(SPHINX_BUILD ${CMAKE_CURRENT_BINARY_DIR}/sphinx)
  #set(SPHINX_INDEX_FILE ${SPHINX_BUILD}/index.html)
  #add_custom_command(
  #  OUTPUT ${SPHINX_INDEX_FILE}
  #  COMMAND 
  #    python3 ../../third-party/chpl-venv/install/chpldeps
  #    sphinx-build -b html
  #    # Tell Breathe where to find the Doxygen output
  #    -Dbreathe_projects.libchplcomp=${DOXYGEN_OUTPUT_DIRECTORY}/xml
  #  ${SPHINX_SOURCE} ${SPHINX_BUILD}
  #  WORKING_DIRECTORY ${CMAKE_CURRENT_BINARY_DIR}
  #  DEPENDS
  #  # Other docs files you want to track should go here (or in some variable)
  #  ${CMAKE_CURRENT_SOURCE_DIR}/index.rst
  #  ${DOXYGEN_INDEX_FILE}
  #  MAIN_DEPENDENCY ${SPHINX_SOURCE}/conf.py
  #  COMMENT "Generating documentation with Sphinx") 
endif(DOXYGEN_FOUND)
