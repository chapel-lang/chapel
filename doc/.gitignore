html
doctrees

/rst/builtins
/rst/modules
/rst/primers
/rst/examples
/rst/language/spec/syntax.rst

man.rst
/rst/tools/chplvis/examples
