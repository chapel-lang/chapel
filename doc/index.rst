.. _chapel-documentation:

Chapel Documentation
====================

.. toctree::
   :caption: Compiling and Running Chapel
   :maxdepth: 1

   Quickstart Instructions <usingchapel/QUICKSTART>
   usingchapel/index
   platforms/index
   technotes/index
   tools/index


.. toctree::
   :caption: Writing Chapel Programs
   :maxdepth: 1

   Quick Reference <language/reference>
   Hello World Variants <examples/index>
   Primers <primers/index>
   Language Specification <language/spec>
   builtins/index
   modules/standard
   modules/packages
   modules/layoutdist
   users-guide/index

.. toctree::
   :caption: Language History
   :maxdepth: 1

   Chapel Evolution <language/evolution>
   Archived Language Specifications <language/archivedSpecs>

Index
-----

* :ref:`Chapel Online Documentation Index <genindex>`

.. COMMENT: clean this up before exposing:: * :chpl:chplref:`chplmodindex`
