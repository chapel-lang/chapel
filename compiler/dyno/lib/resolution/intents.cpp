/*
 * Copyright 2021-2022 Hewlett Packard Enterprise Development LP
 * Other additional copyright holders may be indicated within.
 *
 * The entirety of this work is licensed under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 *
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "chpl/resolution/intents.h"

#include "chpl/types/ClassType.h"
#include "chpl/types/RecordType.h"
#include "chpl/types/UnionType.h"
#include "chpl/types/TupleType.h"

namespace chpl {
namespace resolution {


using namespace uast;
using namespace types;

static QualifiedType::Kind constIntentForType(const Type* t) {

  // anything we don't know the type of has to have unknown intent
  if (t == nullptr || t->isUnknownType() || t->isErroneousType())
    return QualifiedType::UNKNOWN;

  if (t->isPrimitiveType())
    return QualifiedType::CONST_IN;

  if (t->isStringType() || t->isBytesType() ||
      t->isRecordType() || t->isUnionType() || t->isTupleType())
    return QualifiedType::CONST_REF;

  if (auto ct = t->toClassType()) {
    if (ct->decorator().isUnknownManagement())
      return QualifiedType::CONST_INTENT;
    else if (ct->decorator().isManaged())
      return QualifiedType::CONST_REF;
    else
      return QualifiedType::CONST_IN;
  }

  // Otherwise, it should be a generic type that we will
  // instantiate before computing the final intent.
  assert(t->genericity() != Type::CONCRETE);
  return QualifiedType::CONST_INTENT; // leave the intent generic
}

static QualifiedType::Kind defaultIntentForType(const Type* t) {

  // anything we don't know the type of has to have unknown intent
  if (t == nullptr || t->isUnknownType() || t->isErroneousType())
    return QualifiedType::UNKNOWN;

  if (t->isPrimitiveType())
    return QualifiedType::CONST_IN;

  if (t->isStringType() || t->isBytesType() ||
      t->isRecordType() || t->isUnionType() || t->isTupleType())
    return QualifiedType::CONST_REF;

  if (auto ct = t->toClassType()) {
    if (ct->decorator().isUnknownManagement())
      return QualifiedType::DEFAULT_INTENT;
    else if (ct->decorator().isManaged())
      return QualifiedType::CONST_REF;
    else
      return QualifiedType::CONST_IN;
  }

  // Otherwise, it should be a generic type that we will
  // instantiate before computing the final intent.
  assert(t->genericity() != Type::CONCRETE);
  return QualifiedType::DEFAULT_INTENT; // leave the intent generic
}

QualifiedType::Kind resolveIntent(const QualifiedType& t, bool isThis) {
  auto kind = t.kind();
  auto type = t.type();

  // TODO: take into account isThis
  // TODO: REF_MAYBE_CONST or equivalent for arrays, record this

  switch (kind) {
    case QualifiedType::UNKNOWN:
    case QualifiedType::INDEX:
    case QualifiedType::FUNCTION:
    case QualifiedType::MODULE:
      // these don't really have an intent
      return QualifiedType::UNKNOWN;

    case QualifiedType::CONST_REF:
    case QualifiedType::REF:
    case QualifiedType::IN:
    case QualifiedType::CONST_IN:
    case QualifiedType::OUT:
    case QualifiedType::INOUT:
    case QualifiedType::PARAM:
    case QualifiedType::TYPE:
      // concrete intents are already resolved
      return kind;

    case QualifiedType::VAR:
      // normalize VAR to IN to make some test codes easier
      return QualifiedType::IN;
    case QualifiedType::CONST_VAR:
      // normalize CONST_VAR to CONST_IN to make some test codes easier
      return QualifiedType::CONST_IN;

    case QualifiedType::DEFAULT_INTENT:
      // compute the default intent if needed
      return defaultIntentForType(type);

    // compute the const intent if needed
    case QualifiedType::CONST_INTENT:
      return constIntentForType(type);
  }

  return QualifiedType::UNKNOWN;
}


} // end namespace resolution
} // end namespace chpl
