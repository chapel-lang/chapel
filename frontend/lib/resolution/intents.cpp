/*
 * Copyright 2021-2023 Hewlett Packard Enterprise Development LP
 * Other additional copyright holders may be indicated within.
 *
 * The entirety of this work is licensed under the Apache License,
 * Version 2.0 (the "License"); you may not use this file except
 * in compliance with the License.
 *
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "chpl/resolution/intents.h"

#include "chpl/types/ClassType.h"
#include "chpl/types/RecordType.h"
#include "chpl/types/UnionType.h"
#include "chpl/types/TupleType.h"

namespace chpl {
namespace resolution {


using namespace uast;
using namespace types;

static QualifiedType::Kind constIntentForType(const Type* t) {

  // anything we don't know the type of has to have unknown intent
  if (t == nullptr || t->isUnknownType() || t->isErroneousType())
    return QualifiedType::UNKNOWN;

  if (t->isPrimitiveType() || t->isEnumType() || t->isOpaqueType() ||
      t->isTaskIdType()  || t->isNilType() ||
      t->isCStringType() || t->isCVoidPtrType() || t->isCFnPtrType() ||
      t->isNothingType() || t->isVoidType())
    return QualifiedType::CONST_IN;

  if (t->isStringType() || t->isBytesType() ||
      t->isRecordType() || t->isUnionType() || t->isTupleType())
    return QualifiedType::CONST_REF;

  if (auto ct = t->toClassType()) {
    if (ct->decorator().isUnknownManagement())
      return QualifiedType::CONST_INTENT;
    else if (ct->decorator().isManaged())
      return QualifiedType::CONST_REF;
    else
      return QualifiedType::CONST_IN;
  }

  // Otherwise, it should be a generic type that we will
  // instantiate before computing the final intent.
  CHPL_ASSERT(t->genericity() != Type::CONCRETE);
  return QualifiedType::CONST_INTENT; // leave the intent generic
}

static QualifiedType::Kind defaultIntentForType(const Type* t,
                                                bool isThis,
                                                bool isInit) {

  // anything we don't know the type of has to have unknown intent
  if (t == nullptr || t->isUnknownType() || t->isErroneousType())
    return QualifiedType::UNKNOWN;

  if (t->isPrimitiveType() || t->isEnumType() || t->isOpaqueType() ||
      t->isTaskIdType() ||  t->isNilType() ||
      t->isCStringType() || t->isCVoidPtrType() || t->isCFnPtrType() ||
      t->isNothingType() || t->isVoidType())
    return QualifiedType::CONST_IN;

  if (t->isStringType() || t->isBytesType() ||
      t->isRecordType() || t->isUnionType() || t->isTupleType()) {
    if (isThis) {
      if (isInit)
        return QualifiedType::REF;
      else
        return QualifiedType::REF_MAYBE_CONST;
    } else {
      return QualifiedType::CONST_REF;
    }
  }

  if (auto ct = t->toClassType()) {
    if (ct->decorator().isUnknownManagement())
      return QualifiedType::DEFAULT_INTENT;
    else if (ct->decorator().isManaged())
      return QualifiedType::CONST_REF;
    else
      return QualifiedType::CONST_IN;
  }

  // Otherwise, it should be a generic type that we will
  // instantiate before computing the final intent.
  CHPL_ASSERT(t->genericity() != Type::CONCRETE);
  return QualifiedType::DEFAULT_INTENT; // leave the intent generic
}

QualifiedType::Kind resolveIntent(const QualifiedType& t,
                                  bool isThis,
                                  bool isInit) {
  auto kind = t.kind();
  auto type = t.type();

  switch (kind) {
    case QualifiedType::UNKNOWN:
    case QualifiedType::INDEX:
    case QualifiedType::TYPE_QUERY:
    case QualifiedType::PARENLESS_FUNCTION:
    case QualifiedType::FUNCTION:
    case QualifiedType::MODULE:
      // these don't really have an intent
      return QualifiedType::UNKNOWN;

    case QualifiedType::REF_MAYBE_CONST:
      // it's already resolved as much as it wil be here
      return kind;

    case QualifiedType::CONST_REF:
    case QualifiedType::REF:
    case QualifiedType::IN:
    case QualifiedType::CONST_IN:
    case QualifiedType::OUT:
    case QualifiedType::INOUT:
    case QualifiedType::PARAM:
    case QualifiedType::TYPE:
      // concrete intents are already resolved
      return kind;

    case QualifiedType::VAR:
      // normalize VAR to IN to make some test codes easier
      return QualifiedType::IN;
    case QualifiedType::CONST_VAR:
      // normalize CONST_VAR to CONST_IN to make some test codes easier
      return QualifiedType::CONST_IN;

    case QualifiedType::DEFAULT_INTENT:
      // compute the default intent if needed
      return defaultIntentForType(type, isThis, isInit);

    // compute the const intent if needed
    case QualifiedType::CONST_INTENT:
      return constIntentForType(type);
  }

  return QualifiedType::UNKNOWN;
}


} // end namespace resolution
} // end namespace chpl
