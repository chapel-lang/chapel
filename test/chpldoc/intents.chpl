module Inner {
  /* This function has an argument with the in intent */
  proc one (in val) {

  }

  /* This function has an argument with the out intent */
  proc two (out val) {

  }

  /* This function has an argument with the inout intent */
  proc three (inout val) {

  }

  /* This function has an argument with the param intent */
  proc four (param val) {

  }

  /* This function has an argument with the type intent */
  proc five (type val) {

  }

  /* This function has an argument with the const intent */
  proc fiveA (const val) {

  }

  /* This function has an argument with the ref intent */
  proc fiveB (ref val) {
  
  }

  /* This function has an argument with an intent and a type */
  proc six (in val: int) {

  }

  /* This function has a return intent of ref */
  proc seven () ref {
    return "I suppose if I must have a return type here . . . ";
  }

  /* This function has a return intent of const */
  proc eight () const {

  }

  /* This function has a return intent of param */
  proc nine () param {

  }

  /* This function has a return intent of type */
  proc ten () type {

  }

  /* This function has both an argument and a return intent */
  proc twelve (val: int) param {

  }

  /* This function has both an argument intent and a return intent */
  proc thirteen (inout val) param {

  }

  /* This function has a return intent (other than ref) and a body */
  proc fourteen () const {
    return true;
  }
}
/* This function has an argument with the in intent */
proc one (in val) {

}

/* This function has an argument with the out intent */
proc two (out val) {
  
}

/* This function has an argument with the inout intent */
proc three (inout val) {
  
}

/* This function has an argument with the param intent */
proc four (param val) {
  
}

/* This function has an argument with the type intent */
proc five (type val) {
  
}

/* This function has an argument with the const intent */
proc fiveA (const val) {

}

/* This function has an argument with the ref intent */
proc fiveB (ref val) {

} 

/* This function has an argument with an intent and a type */
proc six (in val: int) {
  
}

/* This function has a return intent of ref */
proc seven () ref {
  return "I suppose if I must have a return type here . . . ";
}

/* This function has a return intent of const */
proc eight () const {
  
}

/* This function has a return intent of param */
proc nine () param {
  
}

/* This function has a return intent of type */
proc ten () type {
  
}

/* This function has both an argument and a return intent */
proc twelve (val: int) param {
  
}

/* This function has both an argument intent and a return intent */
proc thirteen (inout val) param {

}

/* This function has a return intent (other than ref) and a body */
proc fourteen () const {
  return true;
}
