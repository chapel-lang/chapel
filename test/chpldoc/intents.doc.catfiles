docs/intents.doc.txt
docs/intents.doc/Inner.txt
