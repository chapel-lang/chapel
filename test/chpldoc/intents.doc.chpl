module Inner {
  /* This function has an argument with the in intent */
  proc one (in val) {

  }

  /* This function has an argument with the out intent */
  proc two (out val) {

  }

  /* This function has an argument with the inout intent */
  proc three (inout val) {

  }

  /* This function has an argument with the param intent */
  proc four (param val) {

  }

  /* This function has an argument with the type intent */
  proc five (type val) {

  }

  /* This function has an argument with the const intent */
  proc fiveA (const val) {

  }

  /* This function has an argument with the ref intent */
  proc fiveB (ref val) {

  }

  /* This function has an argument with an intent and a type */
  proc six (in val: int) {

  }

  /* This function has a return intent of ref */
  proc seven () ref {
    return "I suppose if I must have a return type here . . . ";
  }

  /* This function has a return intent of const */
  proc eight () const {

  }

  /* This function has a return intent of param */
  proc nine () param {

  }

  /* This function has a return intent of type */
  proc ten () type {

  }

  /* This function has both an argument and a return intent */
  proc twelve (val: int) param {

  }

  /* This function has both an argument intent and a return intent */
  proc thirteen (inout val) ref {

  }

  /* This function has a return intent (other than ref) and a body */
  proc fourteen () const {
    return true;
  }
}
/* This function has an argument with the in intent */
proc one (in val) {

}

/* This function has an argument with the out intent */
proc two (out val) {

}

/* This function has an argument with the inout intent */
proc three (inout val) {

}

/* This function has an argument with the param intent */
proc four (param val) {

}

/* This function has an argument with the type intent */
proc five (type val) {

}

/* This function has an argument with the const intent */
proc fiveA (const val) {

}

/* This function has an argument with the ref intent */
proc fiveB (ref val) {

}

/* This function has an argument with an intent and a type */
proc six (in val: int) {

}

/* This function has a return intent of ref */
proc seven () ref {
  return "I suppose if I must have a return type here . . . ";
}

/* This function has a return intent of const */
proc eight () const {

}

/* This function has a return intent of param */
proc nine () param {

}

/* This function has a return intent of type */
proc ten () type {

}

/* This function has both an argument and a return intent */
proc twelve (val: int) param {

}

/* This function has both an argument intent and a return intent */
proc thirteen (inout val) ref {

}

/* This function has a return intent (other than ref) and a body */
proc fourteen () const {
  return true;
}


record foo {
  /* this method has a "const" "this" intent */
  proc const contains(x: eltType): bool {
    return true;
  }

  /*
  This method has a "const ref" "this" intent
  */
  proc const ref eighteen {
    return false;
  }
}

/* I am test fifteen, a test for return intent and type */
proc fifteen(type t) param : int {
  return 1;
}

/* I am test sixteen, a test for return intent and type */
proc sixteen(v: int ) param : bool {
  return true;
}

/* I am test seventeen, cousin of sixteen,
   also a test for return intent and type */
proc seventeen(param q: bool) param : string {
  return "yes";
}
