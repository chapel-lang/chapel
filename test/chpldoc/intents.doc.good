.. default-domain:: chpl

.. module:: intents.doc

intents.doc
===========
**Usage**

.. code-block:: chapel

   use intents.doc;

**Submodules**

.. toctree::
   :maxdepth: 1
   :glob:

   intents.doc/*

.. function:: proc one(in val)

   This function has an argument with the in intent 

.. function:: proc two(out val)

   This function has an argument with the out intent 

.. function:: proc three(inout val)

   This function has an argument with the inout intent 

.. function:: proc four(param val)

   This function has an argument with the param intent 

.. function:: proc five(type val)

   This function has an argument with the type intent 

.. function:: proc fiveA(const val)

   This function has an argument with the const intent 

.. function:: proc fiveB(ref val)

   This function has an argument with the ref intent 

.. function:: proc six(in val: int)

   This function has an argument with an intent and a type 

.. function:: proc seven() ref

   This function has a return intent of ref 

.. function:: proc eight()

   This function has a return intent of const 

.. function:: proc nine() param

   This function has a return intent of param 

.. function:: proc ten() type

   This function has a return intent of type 

.. function:: proc twelve(val: int) param

   This function has both an argument and a return intent 

.. function:: proc thirteen(inout val) param

   This function has both an argument intent and a return intent 

.. function:: proc fourteen()

   This function has a return intent (other than ref) and a body 

.. default-domain:: chpl

.. module:: Inner

Inner
=====
**Usage**

.. code-block:: chapel

   use intents.doc.Inner;

.. function:: proc one(in val)

   This function has an argument with the in intent 

.. function:: proc two(out val)

   This function has an argument with the out intent 

.. function:: proc three(inout val)

   This function has an argument with the inout intent 

.. function:: proc four(param val)

   This function has an argument with the param intent 

.. function:: proc five(type val)

   This function has an argument with the type intent 

.. function:: proc fiveA(const val)

   This function has an argument with the const intent 

.. function:: proc fiveB(ref val)

   This function has an argument with the ref intent 

.. function:: proc six(in val: int)

   This function has an argument with an intent and a type 

.. function:: proc seven() ref

   This function has a return intent of ref 

.. function:: proc eight()

   This function has a return intent of const 

.. function:: proc nine() param

   This function has a return intent of param 

.. function:: proc ten() type

   This function has a return intent of type 

.. function:: proc twelve(val: int) param

   This function has both an argument and a return intent 

.. function:: proc thirteen(inout val) param

   This function has both an argument intent and a return intent 

.. function:: proc fourteen()

   This function has a return intent (other than ref) and a body 

