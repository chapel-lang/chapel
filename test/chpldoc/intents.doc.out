Module: intents
   proc one(in val)
       This function has an argument with the in intent 

   proc two(out val)
       This function has an argument with the out intent 

   proc three(inout val)
       This function has an argument with the inout intent 

   proc four(param val)
       This function has an argument with the param intent 

   proc five(type val)
       This function has an argument with the type intent 

   proc fiveA(const val)
       This function has an argument with the const intent 

   proc fiveB(ref val)
       This function has an argument with the ref intent 

   proc six(in val: int(64))
       This function has an argument with an intent and a type 

   proc seven() ref
       This function has a return intent of ref 

   proc eight()
       This function has a return intent of const 

   proc nine() param
       This function has a return intent of param 

   proc ten() type
       This function has a return intent of type 

   proc twelve(val: int(64)) param
       This function has both an argument and a return intent 

   proc thirteen(inout val) param
       This function has both an argument intent and a return intent 

   proc fourteen()
       This function has a return intent (other than ref) and a body 

   Module: intents.Inner
      proc one(in val)
          This function has an argument with the in intent 

      proc two(out val)
          This function has an argument with the out intent 

      proc three(inout val)
          This function has an argument with the inout intent 

      proc four(param val)
          This function has an argument with the param intent 

      proc five(type val)
          This function has an argument with the type intent 

      proc fiveA(const val)
          This function has an argument with the const intent 

      proc fiveB(ref val)
          This function has an argument with the ref intent 

      proc six(in val: int(64))
          This function has an argument with an intent and a type 

      proc seven() ref
          This function has a return intent of ref 

      proc eight()
          This function has a return intent of const 

      proc nine() param
          This function has a return intent of param 

      proc ten() type
          This function has a return intent of type 

      proc twelve(val: int(64)) param
          This function has both an argument and a return intent 

      proc thirteen(inout val) param
          This function has both an argument intent and a return intent 

      proc fourteen()
          This function has a return intent (other than ref) and a body 

